File: M3Tools/Components/Database/InventoryDatabase.vb
Imports SPPBC.M3Tools.Types
Imports System.Data.SqlClient
Imports System.ComponentModel

Namespace Database
	Public NotInheritable Class InventoryDatabase
		Private Const path As String = "inventory"

		<Description("The username to use for the database connection")>
		<SettingsBindable(True)>
		Public Property Username As String
			Get
				Return dbConnection.Username
			End Get
			Set(value As String)
				dbConnection.Username = value
			End Set
		End Property

		'The password to use for the database connection
		<PasswordPropertyText(True)>
		<SettingsBindable(True)>
		<Description("The password to use for the database connection")>
		Public Property Password As String
			Get
				Return dbConnection.Password
			End Get
			Set(value As String)
				dbConnection.Password = value
			End Set
		End Property

		'The initial catalog to use for the database connection
		<Bindable(True)>
		<Description("The initial catalog to use for the database connection")>
		<SettingsBindable(True)>
		Public Property BaseUrl As String
			Get
				Return dbConnection.BaseUrl
			End Get
			Set(value As String)
				dbConnection.BaseUrl = value
			End Set
		End Property

		Public Function GetProduct(itemID As Integer) As Product
			Return dbConnection.Consume(Of Product)(M3API.Method.Get, $"/{path}/{itemID}").Result
		End Function

		Public Function GetProducts() As DBEntryCollection(Of Product)
			Return dbConnection.Consume(Of DBEntryCollection(Of Product))(Method.Get, $"/{path}").Result
		End Function

		Public Sub AddProduct(itemName As String, stock As Integer, price As Decimal)
			AddProduct(New Product(-1, itemName, stock, price, True))
		End Sub

		Public Sub AddProduct(item As Product)
			dbConnection.Consume(Method.Post, $"/{path}", JSON.ConvertToJSON(item))
		End Sub

		Public Sub UpdateProduct(itemID As Integer, itemName As String, stock As Integer, price As Decimal, available As Boolean)
			UpdateProduct(New Product(itemID, itemName, stock, price, available))
		End Sub

		Public Sub UpdateProduct(item As Product)
			dbConnection.Consume(Method.Put, $"/{path}/{item.Id}", JSON.ConvertToJSON(item))
		End Sub

		Public Sub RemoveProduct(itemID As Integer)
			dbConnection.Consume(Method.Put, $"/{path}/{itemID}")
		End Sub

		Private Sub ChangeAvailability(ParamArray params As SqlParameter())
			Throw New NotImplementedException("ChangeAvailablity")
		End Sub
	End Class
End Namespace

File: M3Tools/Components/Database/InventoryDatabase.cs
using System;
using System.Threading.Tasks;

using SPPBC.M3Tools.M3API;

namespace SPPBC.M3Tools.Database
{
	/// <summary>
	/// The inventory based API calls
	/// </summary>
	public sealed partial class InventoryDatabase
	{
		/// <summary>
		/// Retrieve a product with the specified item ID
		/// </summary>
		/// <param name="itemID"></param>
		/// <param name="ct"></param>
		/// <returns></returns>
		public async Task<Types.Product> GetProduct(int itemID, System.Threading.CancellationToken ct = default)
			=> Utils.ValidID(itemID)
				? await ExecuteWithResultAsync<Types.Product>(System.Net.Http.HttpMethod.Get, string.Join(Paths.Separator, Paths.Inventory, itemID), string.Empty, ct)
				: throw new ArgumentException();

		/// <summary>
		/// Retrieve the complete list of inventory items
		/// </summary>
		/// <returns></returns>
		public async Task<Types.InventoryCollection> GetProducts(System.Threading.CancellationToken ct = default)
			=> await ExecuteWithResultAsync<Types.InventoryCollection>(System.Net.Http.HttpMethod.Get, Paths.Inventory, string.Empty, ct);

		/// <summary>
		/// Add a new product to the database
		/// </summary>
		/// <param name="item"></param>
		/// <param name="ct"></param>
		public async Task<bool> AddInventory(Types.Product item, System.Threading.CancellationToken ct = default)
			=> await ExecuteAsync(System.Net.Http.HttpMethod.Post, Paths.Inventory, JSON.ConvertToJSON(item), ct);

		/// <summary>
		/// 
		/// </summary>
		/// <param name="item"></param>
		/// <param name="ct"></param>
		public async Task<bool> UpdateProduct(Types.Product item, System.Threading.CancellationToken ct = default)
			=> await ExecuteAsync(System.Net.Http.HttpMethod.Put, string.Join(Paths.Separator, Paths.Inventory, item.Id), JSON.ConvertToJSON(item), ct);

		/// <summary>
		/// Remove a product from the database
		/// </summary>
		/// <param name="itemID"></param>
		/// <param name="ct"></param>
		public async void RemoveProduct(int itemID, System.Threading.CancellationToken ct = default)
			=> await ExecuteAsync(System.Net.Http.HttpMethod.Delete, string.Join(Paths.Separator, $"{itemID}?force"), string.Empty, ct);
	}
}
